[Access] Properly handle subscription errors in data providers

illia-malachyn opened the pull request:
Originally, the description of the issue was like that
```
Distinguish between `context.Canceled` errors originating from the streamer and those triggered by the DataProvider’s `Close()` method.

1. I Added `closeChan` that is used in `DataProvider.Close()` method to indicate `DataProvider.Run()` that users of data providers (WebSocket controller in our case) want to finish receiving data. 
2. `HandleSubscription()` function is replaced by `run()` function that is aware of `closeChan`. I made a new function for it because `HandleSubscription()` is widely used in the access package (`HandleRPCSubscription` has 22 usages atm).
3. `run()` returns nil if `closeChan` was closed. `HandleSubscription` returned `ctx.Canceled` which lead to confusion as `ctx.Canceled` could come from 2 sources (streamer and websocket controller).
4. I did a little refactoring to a bunch of `sendResponse()` functions to make it more readable.
```

However, after a couple of discussions and work on that, I can conclude that the first 3 clauses are outdated and were removed from the implementation. The only thing we left is the refactoring and the fact that we start the ubscription after a Run() method has been called.  

Closes #7040 #7047

illia-malachyn commented:
After the discussion with Yurii, I decided to use a different approach here.

peterargue reviewed (commented):
why is this refactor needed? can you point out the important changes, it's not clear to me

illia-malachyn commented:
> why is this refactor needed? can you point out the important changes, it's not clear to me

Hey. I updated PR's description. Added a context of what has been done and why.

illia-malachyn commented on engine/access/rest/websockets/data_providers/account_statuses_provider.go:
```diff
@@ -83,7 +81,48 @@ func NewAccountStatusesDataProvider(
+	// The block counter increments until either:
+	// 1. The account emits events
+	// 2. The heartbeat interval is reached
+	*blocksSinceLastMessage += 1
```
all places with `blockSinceLastMessage` were refactored

illia-malachyn commented on engine/access/rest/websockets/data_providers/account_statuses_provider.go:
```diff
@@ -83,7 +81,48 @@ func NewAccountStatusesDataProvider(
+	}
+
+	var accountStatusesPayload models.AccountStatusesResponse
+	defer messageIndex.Increment()
```
all places with `messageIndex` increment were refactored

illia-malachyn commented on engine/access/rest/websockets/data_providers/base_provider.go:
```diff
@@ -15,6 +17,9 @@ type baseDataProvider struct {
 	send           chan<- interface{}
 	subscription   subscription.Subscription
+	// Ensures the closedChan has been closed once.
+	closedFlag sync.Once
```
Added `closedChan` channel

illia-malachyn commented on engine/access/rest/websockets/data_providers/base_provider.go:
```diff
@@ -56,4 +63,39 @@ func (b *baseDataProvider) Arguments() models.Arguments {
+) error {
+	for {
+		select {
+		case <-closedChan:
```
run() is aware of `closedChan` and returns nil if it was closed

illia-malachyn commented:
@peterargue I also pointed out the most important lines of code

illia-malachyn commented on engine/access/rest/websockets/data_providers/block_digests_provider.go:
```diff
@@ -61,21 +61,23 @@ func NewBlockDigestsDataProvider(
+	return run(
+		p.closedChan,
 		p.subscription,
-		subscription.HandleResponse(p.send, func(b *flow.BlockDigest) (interface{}, error) {
```
Yes. It is used in `HandleRPCSubscription`. The latter is widely used across `backend` package

peterargue commented on engine/access/rest/websockets/data_providers/account_statuses_provider.go:
```diff
@@ -90,7 +88,46 @@ func NewAccountStatusesDataProvider(
+
+	var accountStatusesPayload models.AccountStatusesResponse
+	accountStatusesPayload.Build(response, messageIndex.Value())
+	messageIndex.Increment()
```
this is different from what was there before, though it would only make a difference if `sendResponse` is called concurrently.

Previously, we captured the value, and set `messageIndex` to value + 1. if this failed, we returned an error. Here, we simply increment to some future value, which may or may not be value + 1.

I think this is OK since run will never call `sendResponse` concurrently, but please add a comment noting that the method is not safe for concurrent use.

illia-malachyn replied:
Exactly. This method was not called concurrently but was written in a way it looked like it could be called like that and it was confusing me. The question I asked myself was "If we call it concurrently, why `blockSinceLastMessage` is not an atomic variable?". 

As we call it only from one place, I decided to rewrite it so that it is more readable

illia-malachyn replied:
I'll add an appropriate comment about it

peterargue commented on engine/access/rest/websockets/data_providers/account_statuses_provider.go:
```diff
@@ -90,7 +88,46 @@ func NewAccountStatusesDataProvider(
+	messageIndex := counters.NewMonotonicCounter(0)
+	blocksSinceLastMessage := uint64(0)
+
+	return run(
+		p.closedChan,
+		p.subscription,
+		func(response *backend.AccountStatusesResponse) error {
+			return p.sendResponse(response, &messageIndex, &blocksSinceLastMessage)
+		},
+	)
```
why refactor `handleResponse()`? it seems like this is the same as 
```golang
	return run(p.closedChan, p.subscription, p.handleResponse
```

peterargue commented on engine/access/rest/websockets/data_providers/base_provider.go:
```diff
@@ -56,4 +63,63 @@ func (b *baseDataProvider) Arguments() models.Arguments {
+//   - If a received value is not of the expected type (T), an error is returned.
+//   - If the sendResponse callback encounters an error, it is wrapped and returned.
+func run[T any](
+	closedChan <-chan struct{},
```
we usually call this `done`
```suggestion
	done <-chan struct{},
```

peterargue commented on engine/access/rest/websockets/data_providers/base_provider.go:
```diff
@@ -56,4 +63,63 @@ func (b *baseDataProvider) Arguments() models.Arguments {
 // No errors are expected during normal operations.
 func (b *baseDataProvider) Close() {
 	b.cancel()
+	b.closedFlag.Do(func() {
```
why was this needed? this type of pattern often points to a design problem.

peterargue replied:
could we just use the context's `Done()` channel for `closedChan`?

illia-malachyn replied:
The use of `closedFlag` ensures that the done channel is closed exactly once, even if `Close()` is called multiple times.

We expect clients to call `Run()` and `Close()` from different goroutines—this makes sense because `Run()` is blocking, and clients need to call `Close()` concurrently to stop receiving data.

Since a client might have multiple exit conditions triggering `Close()`, we handle the idempotency internally for simplicity rather than requiring clients to manage it themselves. This prevents potential panics from closing an already-closed channel.

illia-malachyn replied:
Indeed we can use a ctx over a done channel here. However, I'm not sure how to do so correctly without storing ctx in struct. Will ping you once it is ready

peterargue replied:
> Read a PR & linked issue description for it.

@illia-malachyn can you add a pointer to that PR/issue here. 

> In this case, we fix the potential issue and we don't have to worry about it.

which issue are you referring to?

I'm not clear how using a separate done channel is different from a context's done channel wrt lost errors. It seems like it's possible in both cases.

illia-malachyn replied:
I'm not sure the above link was created well. I'm speaking of this place in the `run()` function

```
		case value, ok := <-subscription.Channel():
			if !ok {
				err := subscription.Err()
				if err != nil {
					return fmt.Errorf("subscription finished with error: %w", err)  // who cancelled the context??
				}

				return nil
			}
	```

peterargue replied:
> We cannot know exactly what happened - is it a controller who initiated a shutdown or has something happened in the subscription?

Looking over the code, I think we're over thinking this.

The only way the subscription's error is `context.Canceled` is if the context passed to the streamer was canceled. That context is passed by the data provider, which originally comes from the controller. This means that if the error is context canceled, either:
1. The controller called the provider's `Close()` method.
2. The controller's parent context was cancelled, signalling the node is shutting down.

I think a simple check if the error is context canceled in the conditional should be all that's needed.

I think it's OK for now to handle server shutdowns by gracefully shutting down the connection. Ideally, we'd signal the shutdown with an error to the user, but that can come later. One note from looking over the code, when the controller's context is canceled (server shutdown), we start dropping all messages, so any error response would be dropped anyway:
https://github.com/onflow/flow-go/blob/a149dc5a837656dcc781091063ae2db90d511f72/engine/access/rest/websockets/controller.go#L568-L574 

> We still may run into a race condition if 2 events happen concurrently (controller shutdowns provider and error happening in subscription/streamer)

I don't think it's possible for there to be 2 errors happening simultaneously. The streamer is single threaded and handles each message serially. If the context is canceled, it will detect that before trying processes more data. If the subscription encountered an error at the same time as the context was canceled, the streamer would handle that error before checking the context at the beginning of the loop. If there are any codepaths that violate this, we should fix them.

my proposal is to simply use
```golang
value, ok := <-subscription.Channel()
if !ok {
	err := subscription.Err()
	if err != nil && !errors.Is(err, context.Canceled) {
		return fmt.Errorf("subscription finished with error: %w", err)
	}

	return nil
}
...
```

this removes the need for the extra done channel, and we can simply call cancel in the `Close()` method

illia-malachyn replied:
I added a comment to a `Streamer.Stream()` function that it sets  `ctx.Canceled` to subscription as error by contract. A client may treat it as a graceful shutdown and can rely on that.

peterargue commented on engine/access/rest/websockets/data_providers/events_provider.go:
```diff
@@ -89,42 +89,43 @@ func NewEventsDataProvider(
+	return run(
+		p.closedChan,
+		p.subscription,
+		func(response *backend.EventsResponse) error {
```
same question about refactoring. to me, the previous code was clearer, especially since it mimics the patterns we use in the backend implementations

peterargue commented on engine/access/rest/websockets/data_providers/base_provider.go:
```diff
@@ -48,12 +71,73 @@ func (b *baseDataProvider) Topic() string {
 }
 
 // Close terminates the data provider.
 //
 // No errors are expected during normal operations.
```
not your changes, but no need to define errors when the function doesn't return any
```suggestion
```

illia-malachyn commented on engine/access/rest/websockets/data_providers/block_digests_provider_test.go:
```diff
@@ -47,7 +46,7 @@ func (s *BlockDigestsProviderSuite) TestBlockDigestsDataProvider_HappyPath() {
 
 // validBlockDigestsArgumentsTestCases defines test happy cases for block digests data providers.
-// Each test case specifies input arguments, and setup functions for the mock API used in the test.
+// Each test case specifies input rawArguments, and setup functions for the mock API used in the test.
```
this was done by IDE. it's unintentional

illia-malachyn commented on engine/access/rest/websockets/data_providers/account_statuses_provider.go:
```diff
@@ -25,76 +23,115 @@ type accountStatusesArguments struct {
+	defer cancel()
+	p.subscriptionState = newSubscriptionState(cancel, p.createAndStartSubscription(ctx, p.arguments))
+
+	// set to nils in case Run() called for the second time
+	p.messageIndex = counters.NewMonotonicCounter(0)
+	p.blocksSinceLastMessage = 0
```
So, this is the question of whether we allow clients to reuse a provider after `Run()`/`Close()` pair. With the current code, we allow do so. You wanna restrict it?

peterargue replied:
I think we should not allow it.

illia-malachyn replied:
As discussed at the meeting, we decided to put comment that client must call it once by contract

peterargue commented on engine/access/rest/websockets/data_providers/base_provider.go:
```diff
@@ -48,12 +71,71 @@ func (b *baseDataProvider) Topic() string {
+	b.doneOnce.Do(func() {
+		close(b.done)
+	})
+	b.subscriptionState.cancelSubscriptionContext()
```
from https://github.com/onflow/flow-go/pull/7046#discussion_r1995483041

> We expect clients to call `Run()` and `Close()` from different goroutines—this makes sense because `Run()` is blocking, and clients need to call `Close()` concurrently to stop receiving data.

Since `Close()` is called in a separate gorountine, it's possible for it to be called before `subscriptionState` is set in `Run()`. e.g. if the client disconnects while subscribing. This is likely to cause occasional crashes.

I think we need to go back to passing the context during initialization so we can initialize the cancel before returning the object.

illia-malachyn replied:
Good catch. I thought maybe I could init subscriptionState to some zero values with a noop cancel function but it leads to a data race anyway. We would have to use a mutex in that case and it starts feeling clumsy. So, I'm going back to passing context to constructor

peterargue commented on engine/access/rest/websockets/data_providers/account_statuses_provider.go:
```diff
@@ -44,55 +42,86 @@ func NewAccountStatusesDataProvider(
-	return subscription.HandleSubscription(p.subscription, p.handleResponse())
+	return run(
+		p.createAndStartSubscription(p.ctx, p.arguments),
+		func(response *backend.AccountStatusesResponse) error {
+			return p.sendResponse(response)
+		},
```
```suggestion
		p.sendResponse,
```

peterargue commented on engine/access/rest/websockets/data_providers/events_provider.go:
```diff
@@ -43,89 +43,83 @@ func NewEventsDataProvider(
-	return subscription.HandleSubscription(p.subscription, p.handleResponse())
+	return run(
+		p.createAndStartSubscription(p.ctx, p.arguments),
+		func(response *backend.EventsResponse) error {
+			return p.sendResponse(response)
+		},
```
```suggestion
		p.sendResponse,
```

peterargue commented on engine/access/rest/websockets/data_providers/transaction_statuses_provider.go:
```diff
@@ -43,75 +41,70 @@ func NewTransactionStatusesDataProvider(
-	return p.api.SubscribeTransactionStatuses(ctx, args.TxID, entities.EventEncodingVersion_JSON_CDC_V0)
+	return run(
+		p.createAndStartSubscription(p.ctx, p.arguments),
+		func(response []*access.TransactionResult) error {
+			return p.sendResponse(response)
+		},
```
```suggestion
		p.sendResponse,
```

peterargue commented on engine/access/rest/websockets/data_providers/block_digests_provider.go:
```diff
@@ -31,55 +30,56 @@ func NewBlockDigestsDataProvider(
-		subscription.HandleResponse(p.send, func(b *flow.BlockDigest) (interface{}, error) {
+	return run(
+		p.createAndStartSubscription(p.ctx, p.arguments),
+		func(b *flow.BlockDigest) error {
```
for consistency, can you break this out into a `p.sendResponse` method

peterargue commented on engine/access/rest/websockets/data_providers/block_headers_provider.go:
```diff
@@ -32,57 +31,56 @@ func NewBlockHeadersDataProvider(
-
+	return run(
+		p.createAndStartSubscription(p.ctx, p.arguments),
+		func(header *flow.Header) error {
```
here too

peterargue commented on engine/access/rest/websockets/data_providers/blocks_provider.go:
```diff
@@ -44,63 +42,68 @@ func NewBlocksDataProvider(
-
+	return run(
+		p.createAndStartSubscription(p.ctx, p.arguments),
+		func(block *flow.Block) error {
```
here too

peterargue commented on engine/access/rest/websockets/data_providers/send_and_get_transaction_statuses_provider.go:
```diff
@@ -44,77 +42,72 @@ func NewSendAndGetTransactionStatusesDataProvider(
-	return p.api.SendAndSubscribeTransactionStatuses(ctx, &args.Transaction, entities.EventEncodingVersion_JSON_CDC_V0)
+	return run(
+		p.createAndStartSubscription(p.ctx, p.arguments),
+		func(response []*access.TransactionResult) error {
+			return p.sendResponse(response)
+		},
```
```suggestion
		p.sendResponse,
```

peterargue reviewed (commented):
one small improvement, then this is ready to merge
